Improve tracing + refactor + bug fix

veerajsk opened the pull request:
- setTerminatedState does not work as expected. If init is not called
the terminate job should return but that does not happen
- refactor code so that more common logic can be moved into the
RunnableObject
- improve tracing by introducing different tracing types & improve
logging

veerajsk committed:
Improve tracing + refactor + bug fix

- setTerminatedState does not work as expected. If init is not called
the terminate job should return but that does not happen
- refactor code so that more common logic can be moved into the
RunnableObject
- improve tracing by introducing different tracing types & improve
logging

veerajsk commented:
tracing changes are taken from https://github.com/rapyuta-robotics/alica/pull/180

Dale-Koenig commented on alica_engine/include/engine/RunnableObject.h:
```diff
@@ -30,11 +31,11 @@ class RunnableObject
     void setEngine(AlicaEngine* engine) { _engine = engine; };
     void setConfiguration(const Configuration* conf) { _configuration = conf; };
     void setName(const std::string& name) { _name = name; };
+    std::string getName() { return _name; }
```
Maybe return by const ref

veerajsk replied:
done

abhi1287 committed:
Merge branch 'rr_devel' into vsk_improve_tracing_plus_refactor

Dale-Koenig commented on alica_engine/src/engine/BasicBehaviour.cpp:
```diff
@@ -44,6 +44,18 @@ AgentId BasicBehaviour::getOwnId() const
     return _engine->getTeamManager().getLocalAgentID();
 }
 
+void BasicBehaviour::stop()
```
We need a better naming scheme than this.  There are too many similar named functions that call each other

veerajsk replied:
This is pretty common where the derived class method calls the base class method of the same name. Eg. operator=. I did a small variation by prefixing with`do` just to maintain consistency in the naming with doInit etc.

Dale-Koenig replied:
They don't have the same name here right?  The problem is the large amount of functions with similar names:
`doStart` `doInit` `initializeParameters` `initJob` `start` `activate` `_onInit`

It makes it really hard to remember which functions belong to which class and are called at what time.

veerajsk replied:
unfortunately the names are part of the external interface because of the auto generated code & we wanted to maintain backward compatibility. For behaviour's we have initialiseParameters, run & onTermination, but for plans we have onInit, run & onTerminate. Further behaviour's have an init() public method as well which I have no clue who uses it so I decided to follow the convention: Runnable object uses - doStart, doStop ... doRun etc. & the derived classes use initJob, runJob etc. & onInit_, onRun_ are wrappers for the auto generated code. @abhi1287 what's your take on this?

christianhelm replied:
Maybe just add a comment that explains the purpose of the functions. E.g external interface, wrappers, ...

Dale-Koenig commented on alica_engine/src/engine/BasicBehaviour.cpp:
```diff
@@ -113,30 +100,36 @@ void BasicBehaviour::doTrigger()
+    doTerminate();
+}
+
+void BasicBehaviour::onInit_()
```
Rather than having yet another set of similarly named functions just to wrap the behavior/plan specific calls with tracing logic, I would rather the calls to `onTermination()` etc are not wrapped in an additional layer, and we can have a `traceInitializeParameters` virtual function and similar for the other two as necessary

Dale-Koenig replied:
Or maybe just reverse the "dependency" of the functions

have a `doInitGeneric` that contains the shared logic, then a virtual `doInit` that calls the generic function, then does a little bit of behavior/plan specific logic.  That would be much better imo than this, where we have a generic function that calls an overridden specific function.  Either way I think we could get rid of these three functions.

veerajsk replied:
I couldn't find another way to do this because of the termination code in doTerminate() which is hard to separate cleanly. Anyway the reasoning behind the current implementation is: the BasicBehaviour/BasicPlan objects implement & manipulate (schedule/cancel) init, run & terminate jobs which use the Runnable object's init, run & termination (doInit, doRun & doTerminate) logic but override what happens once the init, run, terminate is actually done i.e. onInit_, onRun_ & onTerminate_

veerajsk replied:
is it okay if I get rid of onInit_ & onRun_ but retain onTerminate_?

Dale-Koenig replied:
Maybe we can at least separate the scheduling functions from the JIT functions.  `start` could be renamed `schedule`

scheduleSomething() is called when scheduling and `doSomething` is what is called when actually doing it.  Then we think of a way to make it more clear from the naming which function calls which or which function is the generic one

veerajsk replied:
its already that way except for the start/stop functions. Please see above comment for an explanation

Dale-Koenig commented on alica_engine/src/engine/RunnableObject.cpp:
```diff
@@ -34,70 +38,125 @@ ThreadSafePlanInterface RunnableObject::getPlanContext() const
+        onTerminate_();
+        _trace.reset();
+    } catch (const std::exception& e) {
+        ALICA_ERROR_MSG("[BasicBehaviour] Exception in Behaviour-TERMINATE of: " << getName() << std::endl << e.what());
```
nit: std::endl is kinda overused and better just to use `\n` unless you want to flush the output buffer at that point for some reason

veerajsk replied:
this is also a copy/paste of the old code. I'll remove the std::endl, don't see any reason for using it here.

Dale-Koenig commented on alica_engine/src/engine/RunnableObject.cpp:
```diff
@@ -34,70 +38,125 @@ ThreadSafePlanInterface RunnableObject::getPlanContext() const
+    switch (_tracingType) {
+    case TracingType::DEFAULT: {
+        auto parent = _execContext.load()->getParent();
+        for (; parent && (!parent->getBasicPlan() || !parent->getBasicPlan()->getTraceContext().has_value()); parent = parent->getParent());
```
How is it possible the parent isn't a basic plan?

veerajsk replied:
This PR is essentially supposed to be cherry pick from https://github.com/rapyuta-robotics/alica/pull/180 so I just copied that code here. But I guess getBasicPlan() can never return nullptr so I'll get rid of it

Dale-Koenig commented on alica_engine/src/engine/RunnableObject.cpp:
```diff
@@ -34,70 +38,125 @@ ThreadSafePlanInterface RunnableObject::getPlanContext() const
+            ALICA_ERROR_MSG("[RunnableObject] Custom tracing type specified, but no getter for the trace context is provided. Switching to default tracing type instead");
+            _tracingType = TracingType::DEFAULT;
+            startTrace();
+            break;
```
nit: safer to return imo in case logic gets added after the switch statement, or just check this case when setting trace type to begin with

veerajsk replied:
I'll do the check while setting the tracing type.

Dale-Koenig reviewed (commented):
Are success/failure of plans/states already logged?

veerajsk commented:
> Are success/failure of plans/states already logged?

yes, this is essentially a cherry pick of https://github.com/rapyuta-robotics/alica/pull/180 but I had to do some refactoring to avoid duplicate code & fix the bug mentioned

veerajsk committed:
Address review comments

- Check if custom trace context getter is specified while setting the
tracing type instead of when starting the tracing of the behaviour/plan.
- Remove unnecessary usage of std::endl
- Return name of the RunnableObject by reference

veerajsk committed:
Merge remote-tracking branch 'origin/vsk_improve_tracing_plus_refactor' into vsk_improve_tracing_plus_refactor

veerajsk commented:
the refactor will be done as part of another task & I'll make another PR that only takes in the traces & fixes the onTermination bug.
